File: generate-equations.js
const shuffleString = (str) => {
  const a = str.split("");
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    const temp = a[i];
    a[i] = a[j];
    a[j] = temp;
  }
  return a.join("");
};

const matched = (str) => {
  let count = 0;
  for (const i of str) {
    if (count < 0) return false;
    if (i == "(") count++;
    if (i == ")") count--;
  }
  return count === 0;
};

const regexFindAll = (str, re) => {
  const hits = [];
  let match = null;
  do {
    match = re.exec(str);
    if (match) {
      hits.push(match[0]);
    }
  } while (match);
  return hits;
};

function* product(arr1, repeat) {
  if (!repeat) {
    yield [];
  } else {
    for (const item of arr1) {
      for (const items of product(arr1, repeat - 1)) {
        yield [item, ...items];
      }
    }
  }
}

const guessLength = 8;
const maxOpCount = 3;
const alphabet = shuffleString("1234567890*/+-()");
const ops = "+-*/";
const digits = "0123456789";
const perms = product(alphabet.split(""), guessLength);
console.log(alphabet);

for (const permutation of perms) {
  const exp = permutation.join("");
  let invalid = false;
  if (exp[0] === "+" || exp[0] === "-") invalid = true;

  /* if (exp.includes("(")) {
    const parantheses = regexFindAll(exp, /\(([^)]+)\)/);
    for (const p of parantheses) {
      let operatorPresent = false;
      if (p.length > 0 && ops.includes(p[0])) invalid = true;
      for (const xyz of p) {
        if (xyz === "(") {
          invalid = true;
          break;
        }
        if (!ops.includes(xyz)) continue;
        operatorPresent = true;
        break;
      }
      if (!operatorPresent || invalid) {
        invalid = true;
        break;
      }
    }
  } */

  // if(exp.includes("(") && !(exp.includes("*") || exp.includes('/')) invalid = true;

  if (!invalid) {
    for (let i = 0; i < exp.length - 1; i++) {
      if (ops.includes(exp[i]) && ops.includes(exp[i + 1])) {
        invalid = true;
        break;
      }
      if (exp[i] === "0" && digits.includes(exp[i + 1])) {
        invalid = true;
        break;
      }
    }
  }
  if (!invalid && !matched(exp)) invalid = true;

  op_count = 0;
  for (const char of exp) {
    if (ops.includes(char)) {
      op_count++;
      if (op_count > maxOpCount) {
        invalid = true;
        break;
      }
    }
  }

  if (!invalid) {
    try {
      const result = eval(exp);
      const expWithoutParentheses = exp.replace("(", "").replace(")", "");
      if (result == eval(expWithoutParentheses)) invalid = true;
      if (result == Math.floor(result) && result < 1000 && result > 0 && !invalid)
        console.log(exp, result);
    } catch (e) {
      continue;
    }
  }
}

File: generate-equations.py
# Rewritten in JS!

'''
import itertools
import warnings
import math
import re
from random import shuffle

warnings.filterwarnings("ignore")

def main():
    guess_length = 8
    max_op_count = 3

    goal = 17

    alphabet = '1234567890*/+-'
    if guess_length == 8:  # hard mode
        alphabet += '()'
    ops = '+-*/'
    digits = '0123456789'


    # Lower and upper bounds for counts of characters in the answer
    alphabet_min_max = {}
    for char in alphabet:
        min = 0
        max = math.inf

        alphabet_min_max[char] = (min, max)

    # Remove symbols we know aren't in the answer
    alphabet = list(alphabet)
    shuffle(alphabet)
    for char in alphabet:
        if alphabet_min_max[char][1] == 0:
            alphabet.remove(char)
    alphabet = "".join(alphabet)

    print(f'\nPossible symbols: {alphabet}')

    perms = itertools.product(*[alphabet] * guess_length)  # permutations with replacement

    print('\n========== POSSIBLE ANSWERS ==========')

    for permutation in perms:
        exp = "".join(permutation)

        invalid = False

    
        if exp[0] == '+' or exp[0] == '-':
            invalid = True

        if '(' in exp:
            parantheses = re.findall('\((.*?)\)', exp)
            for p in parantheses:
                operatorPresent = False
                if len(p) > 0 and p[0] in ops:
                    invalid = True
                for xyz in p:
                    if xyz not in ops:
                        continue
                    operatorPresent = True
                    break
                    if xyz == '(':
                        invalid = True
                        break
                if not operatorPresent or invalid:
                    invalid = True
                    break
        
        if not invalid:
            for i in range(len(exp) - 1):
                # No two operators in a row (not sure if this is right)
                if exp[i] in ops and exp[i + 1] in ops:
                    invalid = True
                    break
                # No leading 0's
                if exp[i] == '0' and exp[i + 1] in digits:
                    invalid = True
                    break

        if not invalid and not matched(exp):
            invalid = True

        # Check number of operators
        op_count = 0
        for char in exp:
            if char in ops:
                op_count += 1
                if op_count > max_op_count:
                    invalid = True
                    break

        if not invalid:
            try:
                result = eval(exp)
                expWithoutParentheses = exp.replace('(','').replace(')','')
                if result == eval(expWithoutParentheses):
                    invalid = True
                if result == goal and not invalid:
                    print(exp)
            except Exception:
                pass


# https://stackoverflow.com/a/38834005/13176711
def matched(str):
    count = 0
    for i in str:
        if i == "(":
            count += 1
        elif i == ")":
            count -= 1
        if count < 0:
            return False
    return count == 0


if __name__ == '__main__':
    main()
'''